### ISSUE TITLE
Code from wrong git repo branch is being loaded when using library with private git repo with branch ref.

### ISSUE THREAD

**Issue by JeffreyMcJeffFace** [dart-lang/pub#3263]
# Environment

- `pub version` or `flutter pub version`: 2.14.4
- OS version: Windows

# Problem

I have an identical setup as described in this issue: https://github.com/dart-lang/pub/issues/2651

I have 2 Dart projects: Project A and Project B.

Project A is a Flutter app
Project B is a package
Project A depends on B via a git reference:

```yaml
  flutter_repo_b:
    git:
      url: https://github.com/company/flutter_repo_a.git
      ref: dev
```

The correct branch code is _not_ being retrieved. Currently, the "main" branch is the only branch I can get to work.

Notably, despite the codebase being incorrect,  the pubspec.lock file shows both the correct branch and resolved-ref values:

![image](https://user-images.githubusercontent.com/13423987/147292969-fe5e37ee-6afd-420e-8125-54d33648676b.png)

If I make a commit to branch _dev_ on Project B and then do a flutter pub get on Project A, the pubspec.log file **will** update with the latest git commit reference, but continue to use the _main_ branch code.

My Project B Main branch currently has a little bit of code missing, and that's how I'm able to confirm that the wrong codebase is being loaded.   If I pull the dev branch into a local repo and reference like this, it is fine:

```yaml
  flutter_repo_b:
    path: ../project_b
```

## Expected behavior

A branch ref in pubspec.yaml should pull the code from designated branch.

## Actual behavior

The designated branch's metadata is being retrieved, but the actual code retrieved is the "main" branch.

**Comment by slagathar** [comment]
I did a ```flutter pub clean``` and then a ```flutter pub get``` but it still loaded the _main_ branch code.

Next, I modified the version number of the _dev_ branch's pubspec.yaml, so that:

branch main: 0.0.2
branch dev: 0.0.20

I deleted pubspec.lock then did a ```flutter pub get``` and received this message:

> Package doesn't exist (the pubspec for company_api_client 0.0.20 from git has version 0.0.2).

**Comment by slagathar** [comment]
I spun up a couple very basic repos with the same format. Whenever I change the ref value for the git repo, the correct branch loads after doing a ```flutter pub upgrade```

But this process does not work for our much larger projects.

I guess I could summarize the issue as this:

What is the correct process to change the branch a package loads?

From this:

```yaml
  tryl_api_client:
    git:
      url: git@github.com:company/company_api_client.git
      ref: main
```

to


```yaml
  tryl_api_client:
    git:
      url: git@github.com:company/company_api_client.git
      ref: dev
```

After you make that pubspec.yaml change, then what steps should I take to ensure the code changes are properly loaded?

**Comment by sigurdm** [comment]
This sounds odd. I tried reproducing this locally but failed.

Try looking in `.dart_tool/package_config.json` to see what directory the code is used from. In that directory should be a git checkout at the correct ref.

Can you make a small replication that I can reproduce this on?

**Comment by slagathar** [comment]
Hi @sigurdm,

I did make a couple basic private repos to try to reproduce this issue, but I could not. Each time I changed the target branch for the dependency, the correct codebase was loaded. 

The project I'm having trouble with is a much larger project with many dependencies. I'm experience the issue right now with this project. The package_config.json does show the correct git references ("dev" branch), but the package code is from the "main" branch reference (I directly committed some comments into the "main" branch for testing).

I'm going to keep trying to modify my demo repos and hopefully find a way to reproduce and update as soon as possible.

**Comment by petrnymsa** [comment]
@slagathar  Any progress? I am experiencing similar issue but with changing refs to "tags", each version is tagged via git tag. 

Only and only if I've run `pub upgrade` twice, the correct version is loaded.

**Comment by slagathar** [comment]
@petrnymsa Unfortunately, no. I'm assuming I must've been dealing with an environment issue as I couldn't reproduce the issue on a different PC. And I can no longer reproduce the issue on the PC that I originally experienced issue on.

**Comment by dawidope** [comment]
Hi @JeffreyMcJeffFace @sigurdm 

I have probably the same issue. I want to add to pubspec.yaml the package from git from specified branch:
```
flutter_markdown:
  git:
    url: https://github.com/GamerHash/flutter_packages.git
    ref: footnote-and-core-decoration-fix
    path: packages/flutter_markdown
```
When I execute pub get I receive:
```
➜ markdown_demo git:(main) flutter pub get
Resolving dependencies... (1.7s)
Package not available (the pubspec for flutter_markdown 0.6.18 from git has version 0.6.17+3).
```
It looks like the version is taken from main branch and then dart cannot find it in chose branch.

You can reproduce it by cloning this demo repo:
https://github.com/dawidope/markdown_demo
